src/Http/Http.Extensions/src/EndpointMetadataContext.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Http;
+
+/// <summary>
+/// Represents the information accessible during endpoint creation by types that implement <see cref="IEndpointMetadataProvider"/>.
+/// </summary>
+public sealed class EndpointMetadataContext
+{
+    /// <summary>
+    /// Creates a new instance of the <see cref="EndpointMetadataContext"/>.
+    /// </summary>
+    /// <param name="method">The <see cref="MethodInfo"/> associated with the current route handler.</param>
+    /// <param name="services">The <see cref="IServiceProvider"/> instance used to access application services.</param>
+    /// <param name="endpointMetadata">The objects that will be added to the metadata of the endpoint.</param>
+    public EndpointMetadataContext(MethodInfo method, IServiceProvider? services, IList<object> endpointMetadata)
```

Comment:
Do you strongly prefer this constructor over init properties? I doubt this will be commonly instantiated by users either way, but I proposed the init properties in the issue so we won't have to add additional constructors later if we decide to add new properties.

Reply:
I started with the `init` properties but then saw that the newly added `RouteHandlerContext` used a constructor so I changed to that. I'm not hugely fussed either way and I see your point. Either way the `EndpointParameterMetadataContext.Parameter` property would have an internal setter to allow for re-use.